File: randomPickWithWeight/randomPickWithWeight.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    Solution(vector<int> w) {
        preSum = w;
        for (int i = 1; i < w.size(); i++) {
            preSum[i] += preSum[i - 1];
        }
        srand(time(0));
    }
    
    int pickIndex() {
        int len = preSum.size();
        if (len == 0)
            return -1;
        int w = rand() % preSum.back() + 1;
        int left = 0, right = len - 1;
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (preSum[mid] < w) {
                left = mid + 1;
            } else if (preSum[mid] > w) {
                right = mid;
            } else {
                return mid;
            }
        }
        return left;
    }
    vector<int> preSum;
};

int main() {
    vector<int> w = {1, 3};
    Solution s(w);
    cout << s.pickIndex() << endl;
    cout << s.pickIndex() << endl;
    cout << s.pickIndex() << endl;
    return 0;
}

File: 528.randomPickWithWeight/randomPickWithWeight.java
class Solution {

    public Solution(int[] w) {
        random = new Random();
        preSum = new int[w.length];
        preSum[0] = w[0];
        for (int i = 1; i < w.length; i++) {
            preSum[i] = preSum[i - 1] + w[i];
        }
    }
    
    public int pickIndex() {
        int len = preSum.length;
        if (len == 1) return 0;
        int n = random.nextInt(preSum[len - 1]) + 1;
        int left = 0, right = len - 1;
        while (left + 1 < right) {
            int mid = left + (right - left) / 2;
            if (preSum[mid] >= n) {
                right = mid;
            } else {
                left = mid;
            }
        }
        if (preSum[left] >= n) return left;
        else if (preSum[right] >= n) return right;
        else return -1;
    }
    
    private int[] preSum;
    Random random;
}

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(w);
 * int param_1 = obj.pickIndex();
 */
